rcall.c

```diff
@@ -906,11 +941,299 @@ SEXP plr_SPI_exec(SEXP rsql) {
+   					offset + sizeof(plcDatatype) * nargs, tx_len);
+   				return NULL;
+   			}
+
+            r_plan->argtypes = malloc(sizeof(plcDatatype) * nargs);
+   			if (r_plan->argtypes == NULL) {
+   				raise_execution_error("Could not allocate %d bytes for argtypes"
+   					" in py_plan", sizeof(plcDatatype) * nargs);
+   				return NULL;
+   			}
+   			memcpy(r_plan->argtypes, start + offset, sizeof(plcDatatype) * nargs);
+   		}
+   	} else {
+   		raise_execution_error("Server returns message type %c, but we expect %c",
+   							  resp->msgtype, MT_RAW);
+   		return NULL;
+   	}
+
+    r_result = R_MakeExternalPtr(r_plan, R_NilValue, R_NilValue);
+
+    free_rawmsg((plcMsgRaw *) resp);
+
+	return r_result;
+}
+
+/*
+ * plr_SPI_execp - The builtin SPI_execp command for the R interpreter
+ */
+SEXP plr_SPI_execp(SEXP rsaved_plan, SEXP rargvalues) {
+    r_saved_plan    *r_plan = (r_saved_plan *) R_ExternalPtrAddr(rsaved_plan);
+    plcArgument     *args;
+	plcConn         *conn = plcconn_global;
+    plcMsgSQL        msg;
+    plcMessage      *resp;
+    plcMsgResult    *result;
+
+	int			     nargs, i, j;
+    int              res = 0;
+
+    SEXP		r_result = NULL,
+                names,
+                row_names,
+                fldvec,
+                obj;
+
+	char		buf[64];
+
+    if (r_plan == NULL) {
+        raise_execution_error("SPI plan does not found");
+        return NULL;
+    }
+
+    nargs = r_plan->nargs;
+    args = pmalloc(sizeof(plcArgument) * nargs);
+	if (nargs > 0)
+	{
+		if (!Rf_isVectorList(rargvalues))
+            raise_execution_error("second parameter must be a list of arguments to the prepared plan");
+
+		if (length(rargvalues) != nargs)
+            raise_execution_error("list of arguments (%d) is not the same length " \
+				  "as that of the prepared plan (%d)",
+				  length(rargvalues), nargs);
+	}
+
+	for (i = 0; i < nargs; i++)
+	{
+		args[i].type.type = r_plan->argtypes[i];
+		args[i].name = NULL; /* We do not need name */
+		args[i].type.nSubTypes = 0;
+		args[i].type.typeName = NULL;
+		args[i].data.value = NULL;
+		
+		PROTECT(obj = VECTOR_ELT(rargvalues, i));
+
+        if(obj != NULL) {
+            args[i].data.isnull = 0;
+            plc_get_output_function(r_plan->argtypes[i])(obj, &args[i].data.value, NULL);
+        } else {
+            /* follow python client */
+            args[i].data.isnull = 1;
+         	args[i].data.value = NULL;
+        }
+
+		UNPROTECT(1);
+	}
+
+    msg.msgtype   = MT_SQL;
+    msg.sqltype   = SQL_TYPE_PEXECUTE;
+	msg.pplan     = r_plan->pplan;
+	msg.limit     = 0;
+	msg.nargs     = nargs;
+	msg.args      = args;
+
+    plcontainer_channel_send(conn, (plcMessage*) &msg);
+	free_arguments(args, nargs, false, false);
+
+receive:
+    res = plcontainer_channel_receive(plcconn_global, &resp);
+    if (res < 0) {
+        raise_execution_error("Error receiving data from the backend, %d", res);
+        return NULL;
+    }
+
```

Comment:
You could extract the duplicated code into one function lik that in python. See receive_from_frontend()